src/components/ImageUpload/ImageUploadProvider.tsx

```diff
@@ -0,0 +1,279 @@
+    size: [],
+    resolution: [],
+    other: [],
+  });
+  const { maxUploadSizeBytes: serverMaxUploadSizeBytes } = useMaxUploadSize();
+  const maxUploadSizeBytes = Math.min(serverMaxUploadSizeBytes, maxImageSize);
+
+  const handleRejectedFile = useCallback((rejectedFile: FileRejection) => {
+    if (
+      rejectedFile.errors.some((error) => error.code === ErrorCode.FileTooLarge)
+    ) {
+      uploadErrors.current.size.push(rejectedFile.file.name);
+    } else {
+      uploadErrors.current.other.push(rejectedFile.file.name);
+    }
+  }, []);
+
+  const uploadImage = useCallback(
+    async (file: File) => {
+      uploadQueue.current.add(file);
+
+      try {
+        const imageData = await file.arrayBuffer();
+        const size = await determineImageSize(imageData);
+
+        if (size.width * size.height > maxResolutionPixels) {
+          throw new ResolutionTooHighError();
+        }
+
+        const uploadResult = await widgetApi.uploadFile(imageData);
+        return { fileName: file.name, size, mxc: uploadResult.content_uri };
+      } catch (error) {
+        console.error('Error uploading an image', error);
+
+        if (error instanceof ResolutionTooHighError) {
+          uploadErrors.current.resolution.push(file.name);
+        } else {
+          uploadErrors.current.other.push(file.name);
+        }
+
+        throw error;
+      } finally {
+        uploadQueue.current.delete(file);
+      }
+    },
+    [widgetApi],
+  );
+
+  const handleDrop = useCallback(
+    async (files: File[], rejectedFiles: FileRejection[]) => {
+      rejectedFiles.forEach(handleRejectedFile);
+
+      if (files.length > 0) {
+        const message = t(
+          'imageUpload.fileIsUploaded',
+          'The file is being uploaded…',
+          {
+            count: files.length,
+          },
+        );
+
+        showSnackbar({
+          key: message,
+          message,
+        });
+      }
+
+      const results = Promise.allSettled(files.map(uploadImage));
+
+      results
+        .then(() => {
+          if (uploadQueue.current.size === 0) {
+            // All uploads finished
+            clearSnackbar();
+            showErrorSnackbars(
+              showSnackbar,
+              uploadErrors.current,
+              t,
+              maxUploadSizeBytes,
+            );
+            uploadErrors.current = { size: [], resolution: [], other: [] };
+          }
+          return;
+        })
+        .catch(() => {
+          // ignore
+        });
+
+      return results;
+    },
+    [
+      clearSnackbar,
+      handleRejectedFile,
+      maxUploadSizeBytes,
+      showSnackbar,
+      uploadImage,
+      t,
+    ],
+  );
+
+  return (
+    <ImageUploadContext.Provider value={{ handleDrop, maxUploadSizeBytes }}>
+      {children}
+    </ImageUploadContext.Provider>
+  );
+}
+
+/**
+ * Show snackbars for the different types of errors.
+ *
+ * @param showSnackbar - Function to show a snackbar {@link SnackbarContextValue}
```

Comment:
Links to `SnackbarContextValue`, is it correct?

Reply:
More specifically it should be `SnackbarContextValue.showSnackbar` (or `SnackbarState.showSnackbar` after rename). Will update it.

Reply:
Done